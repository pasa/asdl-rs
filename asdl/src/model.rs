#[derive(Debug)]
pub struct Asdl {
    pub types: Vec<Type>,
    pub comments: Vec<String>,
}

impl Asdl {
    pub fn get_type_by_name(&self, name: &str) -> Option<&Type> {
        self.types.iter().find(|t| t.id() == name)
    }
}

#[derive(Debug)]
pub enum Type {
    SumType(SumType),
    ProdType(ProdType),
}

impl Type {
    fn id(&self) -> String {
        match self {
            Type::SumType(sty) => sty.id.clone(),
            Type::ProdType(pty) => pty.id.clone(),
        }
    }
}

impl From<SumType> for Type {
    fn from(n: SumType) -> Type {
        Type::SumType(n)
    }
}
impl From<ProdType> for Type {
    fn from(n: ProdType) -> Type {
        Type::ProdType(n)
    }
}

#[derive(Debug)]
pub struct SumType {
    pub id: String,
    pub constructors: Vec<Constructor>,
    pub attributes: Vec<Field>,
    pub comments: Vec<String>,
}

#[derive(Debug)]
pub struct Constructor {
    pub id: String,
    pub fields: Vec<Field>,
    pub comments: Vec<String>,
}

#[derive(Debug)]
pub struct ProdType {
    pub id: String,
    pub fields: Vec<Field>,
    pub comments: Vec<String>,
}

#[derive(Debug)]
pub struct Field {
    pub id: String,
    pub type_id: String,
    pub arity: Arity,
}

#[derive(Debug)]
pub enum Arity {
    Optional,
    Required,
    Repeated,
}
