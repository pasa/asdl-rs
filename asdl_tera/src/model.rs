use std::collections::HashMap;
use serde::Serialize;
use asdl;
use super::Result;

#[derive(Serialize, Debug)]
pub struct Asdl {
    pub types: HashMap<String, Type>,
    pub prod_types: Vec<String>,
    pub sum_types: Vec<String>,
    pub comments: Vec<String>,
}

impl Asdl {
    pub fn parse(asdl: &str) -> Result<Asdl> {
        Ok(Asdl::new(asdl::Asdl::parse(asdl)?))
    }
}

#[derive(Serialize, Debug)]
#[serde(untagged)]
pub enum Type {
    SumType(SumType),
    ProdType(ProdType),
}

impl From<SumType> for Type {
    fn from(n: SumType) -> Type {
        Type::SumType(n)
    }
}
impl From<ProdType> for Type {
    fn from(n: ProdType) -> Type {
        Type::ProdType(n)
    }
}

#[derive(Serialize, Debug)]
pub struct SumType {
    pub id: String,
    pub constructors: Vec<Constructor>,
    pub attributes: Vec<Field>,
    pub is_prod_type: bool, //always false
    pub comments: Vec<String>,
}

#[derive(Serialize, Debug)]
pub struct Constructor {
    pub id: String,
    pub fields: Vec<Field>,
    pub comments: Vec<String>,
}

#[derive(Serialize, Debug)]
pub struct ProdType {
    pub id: String,
    pub fields: Vec<Field>,
    pub is_prod_type: bool, //always true
    pub comments: Vec<String>,
}

#[derive(Serialize, Debug)]
pub struct Field {
    pub id: String,
    pub type_id: String,
    pub is_required: bool,
    pub is_optional: bool,
    pub is_repeated: bool,
}
