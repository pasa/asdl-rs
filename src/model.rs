use std::fmt::Write;

use rustc_hash::FxHashMap;
use heck::MixedCase;
use linked_hash_set::LinkedHashSet;
use serde::Serialize;

use crate::parser;

#[derive(Serialize, Debug)]
pub struct Asdl {
    sum_types: Vec<SumType>,
    prod_types: Vec<ProdType>,
}

#[derive(Serialize, Debug)]
pub(crate) struct SumType {
    id: String,
    constructors: Vec<Constructor>,
    attributes: Vec<Field>,
}

#[derive(Serialize, Debug)]
pub(crate) struct Constructor {
    id: String,
    fields: Vec<Field>,
}

#[derive(Serialize, Debug, Clone, Hash, Eq, PartialEq)]
pub(crate) struct ProdType {
    id: String,
    fields: Vec<Field>,
}

#[derive(Serialize, Debug, Clone, Hash, Eq, PartialEq)]
pub(crate) struct Field {
    id: String,
    type_id: String,
    is_single: bool,
    is_option: bool,
    is_sequence: bool,
}

impl Field {
    fn single(id: String, type_id: String) -> Field {
        Field { id, type_id, is_single: true, is_option: false, is_sequence: false }
    }

    fn option(id: String, type_id: String) -> Field {
        Field { id, type_id, is_single: false, is_option: true, is_sequence: false }
    }

    fn sequence(id: String, type_id: String) -> Field {
        Field { id, type_id, is_single: false, is_option: false, is_sequence: true }
    }
}

impl Asdl {
    pub(crate) fn new(root: &parser::Root) -> Self {
        let mut res = Asdl { prod_types: vec![], sum_types: vec![] };
        for d in root.types() {
            match d.kind() {
                parser::TypeKind::SumType(t) => {
                    res.sum_types.push(sum_type(t));
                }
                parser::TypeKind::ProdType(t) => {
                    res.prod_types.push(prod_type(t));
                }
            }
        }
        res
    }
}

fn sum_type(node: &parser::SumType) -> SumType {
    let id = node.type_id().text().to_string();
    let constructors = node.constructors().map(constructor).collect();
    let attributes = if let Some(attrs) = node.attrs() {
        let mut names = FieldNames::default();
        attrs.fields().map(|f| field(f, &mut names)).collect()
    } else {
        vec![]
    };
    SumType { id, constructors, attributes }
}

fn constructor(node: &parser::Constr) -> Constructor {
    let id = node.id().text().to_string();
    let mut names = FieldNames::default();
    let fields: Vec<Field> = node.fields().map(|f| field(f, &mut names)).collect();
    Constructor { id, fields }
}

fn field(node: &parser::Field, names: &mut FieldNames) -> Field {
    match node.kind() {
        parser::FieldKind::Single(f) => Field::single(
            names.get_or_generate(f.id(), f.type_id()),
            f.type_id().text().to_string(),
        ),
        parser::FieldKind::Opt(f) => Field::option(
            names.get_or_generate(f.id(), f.type_id()),
            f.type_id().text().to_string(),
        ),
        parser::FieldKind::Sequence(f) => Field::sequence(
            names.get_or_generate(f.id(), f.type_id()),
            f.type_id().text().to_string(),
        ),
    }
}

fn prod_type(node: &parser::ProdType) -> ProdType {
    let id = node.type_id().text().to_string();
    let mut names = FieldNames::default();
    let fields: Vec<Field> = node.fields().map(|f| field(f, &mut names)).collect();
    ProdType { id, fields }
}

#[derive(Default)]
struct FieldNames {
    names_indexes: FxHashMap<String, u32>,
}

impl FieldNames {
    fn get_or_generate(&mut self, id: Option<&parser::Id>, type_id: &parser::TypeId) -> String {
        match id {
            Option::Some(id) => id.text().to_string(),
            Option::None => {
                let type_id = type_id.text();
                let index = self.names_indexes.entry(type_id.to_string()).or_insert(0);
                let res = if *index == 0 {
                    type_id.to_string()
                } else {
                    let mut buf = String::new();
                    write!(&mut buf, "{}{}", type_id, index).unwrap();
                    buf
                };
                *index += 1;
                res
            }
        }
    }
}
