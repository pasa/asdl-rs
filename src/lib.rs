mod model;
mod parser;

use tera::*;
use crate::model::*;
use failure::format_err;
use heck::{CamelCase, ShoutySnakeCase, SnakeCase, MixedCase};

pub fn generate(asdl: &str, template: &str) -> String {
    let root: &parser::Root = &parser::parse(&asdl);
    let model = Asdl::from(root);
    let mut tera = Tera::default();
    tera.register_filter("camel", |arg, _| Ok(arg.as_str().unwrap().to_camel_case().into()));
    tera.register_filter("snake", |arg, _| Ok(arg.as_str().unwrap().to_snake_case().into()));
    tera.register_filter("mixed", |arg, _| Ok(arg.as_str().unwrap().to_mixed_case().into()));
    tera.register_filter("SCREAM", |arg, _| {
        Ok(arg.as_str().unwrap().to_shouty_snake_case().into())
    });
    tera.add_raw_template("_src", &template)
        .map_err(|e| format_err!("template parsing error: {:?}", e))
        .unwrap();
    tera.render("_src", &model)
        .map_err(|e| format_err!("template rendering error: {:?}", e))
        .unwrap()
}

#[cfg(test)]
mod tests {
    use crate::*;
    use insta::assert_snapshot_matches;
    use insta::assert_debug_snapshot_matches;

    #[test]
    fn simple_successful_test() {
        let asdl = r"
            stm = Compound(stm s1, stm* s2)
                | Single(stm)
            noFileds = One | Two | Tree
            prodType = (noFileds? f, stm s1)
            ";
        let root: &parser::Root = &parser::parse(&asdl);
        assert_snapshot_matches!("simple_successful_test_syntax", root.debug_dump());
        let model = Asdl::from(root);
        assert_debug_snapshot_matches!("simple_successful_test_model", model)
    }
}
