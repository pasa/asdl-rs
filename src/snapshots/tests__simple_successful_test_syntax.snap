---
created: "2019-07-02T09:34:16.407811459Z"
creator: insta@0.8.1
source: src/lib.rs
expression: root.debug_dump()
---
node Root[0; 173)
  token WS [0; 13)
  node SumType[13; 87)
    node TypeId[13; 16)
      token `stm` [13; 16)
    token WS [16; 17)
    token `=` [17; 18)
    token WS [18; 19)
    node Constr[19; 44)
      node ConstrId[19; 27)
        token `Compound` [19; 27)
      token `(` [27; 28)
      node Single[28; 34)
        node TypeId[28; 31)
          token `stm` [28; 31)
        token WS [31; 32)
        node Id[32; 34)
          token `s1` [32; 34)
      token `,` [34; 35)
      token WS [35; 36)
      node Sequence[36; 43)
        node TypeId[36; 39)
          token `stm` [36; 39)
        token `*` [39; 40)
        token WS [40; 41)
        node Id[41; 43)
          token `s2` [41; 43)
      token `)` [43; 44)
    token WS [44; 61)
    token `|` [61; 62)
    token WS [62; 63)
    node Constr[63; 74)
      node ConstrId[63; 69)
        token `Single` [63; 69)
      token `(` [69; 70)
      node Single[70; 73)
        node TypeId[70; 73)
          token `stm` [70; 73)
      token `)` [73; 74)
    token WS [74; 87)
  node SumType[87; 128)
    node TypeId[87; 95)
      token `noFileds` [87; 95)
    token WS [95; 96)
    token `=` [96; 97)
    token WS [97; 98)
    node Constr[98; 102)
      node ConstrId[98; 101)
        token `One` [98; 101)
      token WS [101; 102)
    token `|` [102; 103)
    token WS [103; 104)
    node Constr[104; 108)
      node ConstrId[104; 107)
        token `Two` [104; 107)
      token WS [107; 108)
    token `|` [108; 109)
    token WS [109; 110)
    node Constr[110; 128)
      node ConstrId[110; 115)
        token `Three` [110; 115)
      token WS [115; 128)
  node ProdType[128; 160)
    node TypeId[128; 136)
      token `prodType` [128; 136)
    token WS [136; 137)
    token `=` [137; 138)
    token WS [138; 139)
    token `(` [139; 140)
    node Opt[140; 151)
      node TypeId[140; 148)
        token `noFileds` [140; 148)
      token `?` [148; 149)
      token WS [149; 150)
      node Id[150; 151)
        token `f` [150; 151)
    token `,` [151; 152)
    token WS [152; 153)
    node Single[153; 159)
      node TypeId[153; 156)
        token `stm` [153; 156)
      token WS [156; 157)
      node Id[157; 159)
        token `s1` [157; 159)
    token `)` [159; 160)
  token WS [160; 173)

